test/unit/types/eval_const.spec.ts

```diff
@@ -1320,6 +1320,191 @@ const cases: Array<[string, (factory: ASTNodeFactory) => Expression, boolean, Va
             true,
             BigInt("0xe8")
         ],
+        /**
+         * @todo It seems that Solidity allows to skip right bytes on following two cases.
+         * This would need a standalone fix in castable(), InferType.inferCommonType(), also in evalBinary().
+         */
+        // [
+        //     "Edge-case <0.5.0: bytes2(0xff00) == byte(0xff)",
+        //     (factory: ASTNodeFactory) =>
+        //         factory.makeBinaryOperation(
+        //             "<missing>",
+        //             "==",
+        //             factory.makeFunctionCall(
+        //                 "bytes2",
+        //                 FunctionCallKind.TypeConversion,
+        //                 factory.makeElementaryTypeNameExpression("type(bytes2)", "bytes2"),
+        //                 [factory.makeLiteral("<missing>", LiteralKind.Number, "ff00", "0xff00")]
+        //             ),
+        //             factory.makeFunctionCall(
+        //                 "byte",
+        //                 FunctionCallKind.TypeConversion,
+        //                 factory.makeElementaryTypeNameExpression("type(byte)", "byte"),
+        //                 [factory.makeLiteral("<missing>", LiteralKind.Number, "ff", "0xff")]
+        //             )
+        //         ),
+        //     true,
+        //     true
+        // ],
+        // [
+        //     "Edge-case <0.5.0: bytes2(0xcc00) < byte(0xee)",
+        //     (factory: ASTNodeFactory) =>
+        //         factory.makeBinaryOperation(
+        //             "<missing>",
+        //             "<",
+        //             factory.makeFunctionCall(
+        //                 "bytes2",
+        //                 FunctionCallKind.TypeConversion,
+        //                 factory.makeElementaryTypeNameExpression("type(bytes2)", "bytes2"),
+        //                 [factory.makeLiteral("<missing>", LiteralKind.Number, "cc00", "0xcc00")]
+        //             ),
+        //             factory.makeFunctionCall(
+        //                 "byte",
+        //                 FunctionCallKind.TypeConversion,
+        //                 factory.makeElementaryTypeNameExpression("type(byte)", "byte"),
+        //                 [factory.makeLiteral("<missing>", LiteralKind.Number, "ee", "0xee")]
+        //             )
+        //         ),
+        //     true,
+        //     true
+        // ],
```

Comment:
@cd1m0 This is actually interesting and I might need help here.

Currently we apply `clampIntToType()` when type of `BinaryOperation` is `IntType`. However in this case we have `BoolType`, so result is not clamped. Other thing here is that two operands are casted to a common type, that is smaller, than type, that is computed by `InferType.inferCommonType()`, that is relying on `castable()`. In `castable()` we have following entry: https://github.com/Consensys/solc-typed-ast/blob/d4dd707aeb1ecff2627893e26e4c6ae1d6fce286/src/types/utils.ts#L431-L434

But in these cases:
```solidity
bytes2(0xcc00) < byte(0xee);
bytes2(0xff00) == byte(0xff);
```
it seems that result is coerced to a smaller size and this is fine even in latest compilers. I tried following:
```typescript
export function evalBinary(node: BinaryOperation, inference: InferType): Value {
    try {
        const leftT = inference.typeOf(node.vLeftExpression);
        const rightT = inference.typeOf(node.vRightExpression);
        const commonT = inference.inferCommonType(leftT, rightT);

        let left = evalConstantExpr(node.vLeftExpression, inference);
        let right = evalConstantExpr(node.vRightExpression, inference);

        if (commonT instanceof FixedBytesType) {
            left = clampIntToType(toInt(left), fixedBytesTypeToIntType(commonT));
            right = clampIntToType(toInt(right), fixedBytesTypeToIntType(commonT));
        }

        const res = evalBinaryImpl(node.operator, left, right);

        if (leftT instanceof NumericLiteralType && rightT instanceof NumericLiteralType) {
            return res;
        }

        const resT = inference.typeOfBinaryOperation(node);

        if (resT instanceof IntType && typeof res === "bigint") {
            return clampIntToType(res, resT);
        }

        return res;
    } catch (e: unknown) {
        if (e instanceof EvalError && e.expr === undefined) {
            e.expr = node;
        }

        throw e;
    }
}
```
But it solves only one case of two. Would you be able to find some time and take a look?

Reply:
Addressed in d22f7d6